module/zstd/zstd.c

```diff
@@ -41,6 +41,19 @@
 MALLOC_DEFINE(M_ZSTD, "zstd", "ZSTD Compressor");
 #endif
 
+/*
+ * this specifies a incremental version of the zstd format
+ * increase it on any new added zstd version.
+ * this is made to ensure data integrity compatiblity since
+ * new versions might enhance the compression algorithm
+ * in a way the data itself will change.
+ */
+#define	ZSTD_VERSION		0
+#define	ZSTD_VERSION_SHIFT	16
+#define	ZSTD_VERSION_MASK	0xffff0000
+#define	ZSTD_LEVEL_SHIFT	0
```

Comment:
drop dead code

Reply:
Github *sigh* ... drop `ZSTD_LEVEL_SHIFT`

Reply:
you can drop it, but its better for description and there are linux macros for bit operations which are using postfixes like MASK and SHIFT for operation. its very common to specify masked values in that way

Reply:
but we don't use them

Reply:
@c0d3z3r0i like it that way. its clear how its encoded at first view and its very common todo so. of course i could now add some SET_VERSION / GET_VERSION macros just to make use of it.

Reply:
I think you guys both have a point:
- Having them makes it more clear
- It's quite common in Linux to include them
- It isn't used nor needed
- It isn't clear this is "non functioning code".

I suggest removing them and replacing them with a short comment with the same meaning. As OpenZFS is going to serve many more operating systems than just linux, which devs might get confused with this non-functioning code. I think a comment with the same meaning would be fine. Something like:

```
# LEVEL and VERSION share the cookie with the following specifications:
# ZSTD_LEVEL = shift 0, mask 0x0000ffff
# ZSTD_VERSION = shift 16, mask 0xffff0000
```

I think thats even more clear than  just adding the code variant of a comment.

Reply:
Well, a comment "level is stored in the lower 16 bits, zstd version in the upper ones" should be more than clear IMHO

Reply:
in latest revision i added set_level / set_version functions which make use of it. it not final yet. just waiting for tests

Reply:
@c0d3z3r0 the comment is placed at set_zstd_level call which describes how it is stored. of course we can duplicate that.

Reply:
@c0d3z3r0 Funny enough I wanted to ask you the same question an hour ago :P

But yeah, I think Behlendorf wants a simple clean solution, not all sorts of new functions.
KISS, just store it cleanly and read when needed (which currently is "not at all").

Reply:
@BrainSlayer I think a somewhat duplicate of that explaination might be nice to have here, yes.